### PR TITLE
Handle conflicting multi-char auto closing pairs

### LINKED ISSUE #72177: auto closing pairs with conflicting patterns problems
```
Version: 1.33.1 (user setup)
Commit: 51b0b28134d51361cf996d2f0a1c698247aeabd8
Date: 2019-04-11T08:27:14.102Z
Electron: 3.1.6
Chrome: 66.0.3359.181
Node.js: 10.2.0
V8: 6.6.346.32
OS: Windows_NT x64 10.0.17763
```

Steps to Reproduce:

1. Create two auto closing pairs in a language configuration file,
```JSON
	"autoClosingPairs": [
		{"open": "(", "close": ")"},
		{"open": "(*", "close": "*)",  "notIn": ["string"]},
	],
```
2. trying using the two character auto closing pair, `(*` and you will get `(**))`.

On the other hand, if you remove the ending ')' from the closing '*)' you almost get normal function, except that in cases where `(` doesn't auto close with `)`, you get `(**`.

Note this condition exists in #57838, in a reference to the Structured Text Language, though it is not shown in the example on that feature request.  Reference the repository https://github.com/Serhioromano/vscode-st for some example.

I think the Auto Closing logic needs to consider when auto closing pairs might conflict with each other.  In this case, '(**)' overlaps with '()'.

### PR DESCRIPTION
**Pull request by olivercoad**
This PR fixes #72177

When autoclosing multi-character pairs, rather than just adding all of `pair.close`, it now only adds the part that hasn't already been autoclosed.  
It does this by finding the longest pair that is contained in the current pair. It then checks if `subAutoClosingPair.close` is already present (it may not necessarily have just been autoclosed) for all selections, and if so only adds a substring of `pair.close` excluding `subAutoClosingPair.close`.

In order to get auto close working when the sub-pair is multiple characters, `_isBeforeClosingBrace` had to be updated to work with multi-character pairs.

And example of where this is the case is the F# pairs
```
"autoClosingPairs": [
...
["<@", "@>"],
["<@@", "@@>"],
...
]
```

In the process of updating `_isBeforeClosingBrace`, I changed it from excluding symmetric pairs to explicitely returning false for all potential starting braces.
I believe the intended purpose of the symmetric pair check was to allow autoclosing before close braces but to not get mixed up when the close brace might actually be an open brace (see #20801).  
However, the previous implementation also allowed autoclosing before starting braces if the autoclose brace was also symmetric. This seems contrary to the method name so I'm assuming it was unintentional (or otherwise it should probably be more explicit);
this change required a couple of minor changes to some unit tests.

### CODE REVIEW

File: src/vs/editor/common/controller/cursorTypeOperations.ts [hunk 1]
```diff
@@ -498,31 +498,30 @@ export class TypeOperations {
...
+			}
 		}
 
-		const thisBraceIsSymmetric = TypeOperations._autoClosingPairIsSymmetric(autoClosingPair);
-		for (const otherAutoClosingPair of otherAutoClosingPairs) {
-			const otherBraceIsSymmetric = TypeOperations._autoClosingPairIsSymmetric(otherAutoClosingPair);
-			if (!thisBraceIsSymmetric && otherBraceIsSymmetric) {
-				continue;
+		// check that lineAfter DOES immediately starts with a closing brace
+		const potentialClosingBraces = config.autoClosingPairs.autoClosingPairsCloseByStart.get(nextChar) || [];
```
Comment (rheh) [inline src/vs/editor/common/controller/cursorTypeOperations.ts:514]: Lines 507-510 almost duplicate these lines.  One could move to a help function and pass in the potential array and start character to check for.

File: src/vs/editor/common/controller/cursorTypeOperations.ts [hunk 2]
```diff
@@ -548,7 +547,26 @@ export class TypeOperations {
...
+	private static _findSubAutoClosingPairClose(config: CursorConfiguration, autoClosingPair: StandardAutoClosingPairConditional): string {
+		// remove parts from largest sub-pair
+		let subPair: StandardAutoClosingPairConditional | null = null;
+		if (autoClosingPair && autoClosingPair.open.length > 1) {
+			let lastChar = autoClosingPair.close.slice(-1);
+			// get candidates with the same last character as close
+			let subPairCandidates = config.autoClosingPairs.autoClosingPairsCloseByEnd.get(lastChar) || [];
+			for (const subPairCandidate of subPairCandidates) {
+				if (subPair === null || subPairCandidate.open.length > subPair.open.length) {
+					if (autoClosingPair.open !== subPairCandidate.open && autoClosingPair.open.includes(subPairCandidate.open) && autoClosingPair.close.endsWith(subPairCandidate.close)) {
```
Comment (rheh) [inline src/vs/editor/common/controller/cursorTypeOperations.ts:559]: Suggestion: move this into a function? 
`function is<Add Descriptive Name Here> (autoClosingPair, subPairCandidate) {
    return (autoClosingPair.open !== subPairCandidate.open && autoClosingPair.open.includes(subPairCandidate.open) && autoClosingPair.close.endsWith(subPairCandidate.close));
}
`

We could then use this to improve readability?

`If (is<Add Descriptive Name Here>(....) {
  subPair = subPairCandidate;
}
`

I only suggest the above, as I have no idea what the if statement is checking for.  The strongly named function would help future authors to understand the intent.

File: src/vs/editor/common/controller/cursorTypeOperations.ts (hunk 1 above)
Comment (olivercoad) [reply src/vs/editor/common/controller/cursorTypeOperations.ts:514]: True, though a helper function would also need a bool flag for checking open/close.

I just checked and it turns out javascript does actually have built in `Array.prototype.some`, which is all it's really doing anyway.

How would you feel about something like this?

```
private static _isBeforeClosingBrace(config: CursorConfiguration, lineAfter: string) {
    // If the start of lineAfter can be interpretted as both a starting or ending brace, default to returning false
    const nextChar = lineAfter.charAt(0);
    const potentialStartingBraces = config.autoClosingPairs.autoClosingPairsOpenByStart.get(nextChar) || [];
    const potentialClosingBraces = config.autoClosingPairs.autoClosingPairsCloseByStart.get(nextChar) || [];

    const isBeforeStartingBrace = potentialStartingBraces.some(x => lineAfter.startsWith(x.open));
    const isBeforeClosingBrace = potentialClosingBraces.some(x => lineAfter.startsWith(x.close));

    return !isBeforeStartingBrace && isBeforeClosingBrace;
}
```

Reviewing that made me realise L507 should technically be autoClosingPairs**Open**ByStart

File: src/vs/editor/common/controller/cursorTypeOperations.ts (hunk 2 above)
Comment (olivercoad) [reply src/vs/editor/common/controller/cursorTypeOperations.ts:559]: Yes a descriptive name is a good idea. Do you think it's necessary to move into it's own function or just extract as a local variable?

It's basically checking that `subPairCandidate` actually is a subPair of `autoClosingPair`. That is to say, `autoClosingPair` strictly contains open and close and that `autoClosingPair` is only adding characters to the start of close and not the end. TBH I can't think of any good names for it other than "isActuallySubPair".

File: src/vs/editor/common/controller/cursorTypeOperations.ts (hunk 1 above)
Comment (rheh) [reply src/vs/editor/common/controller/cursorTypeOperations.ts:514]: This looks good to me.

File: src/vs/editor/common/controller/cursorTypeOperations.ts (hunk 2 above)
Comment (rheh) [reply src/vs/editor/common/controller/cursorTypeOperations.ts:559]: isActuallySubPair is a good name.  Helps me and other future readers to understand what the intent is.

File: src/vs/editor/common/controller/cursorTypeOperations.ts (hunk 1 above)
Comment (olivercoad) [reply src/vs/editor/common/controller/cursorTypeOperations.ts]: Thanks for the review :)

**Comment by alexdima** [comment]
Thank you!

**Pull request merged**